### PR TITLE
[Bugfix] Fix inverted condition causing thinking_token_budget to be silently ignored

### LINKED ISSUE #41672: [Bug]: thinking_token_budget silently ignored when no penalties are set (inverted condition in gpu_input_batch.py)
## Summary

`thinking_token_budget` has no effect for any request that does not also set penalty parameters (`frequency_penalty`, `presence_penalty`, `repetition_penalty`) or `bad_words`. This covers the vast majority of real-world usage.

## Root Cause

Introduced in commit `68dd7db81` (ThinkingBudgetStateHolder, April 29), `gpu_input_batch.py` contains an inverted boolean condition when computing `needs_output_token_ids`:

```python
# BEFORE (buggy) - vllm/v1/worker/gpu_input_batch.py ~line 879
needs_output_token_ids = (
    not self.no_penalties
    or bool(self.bad_words_token_ids)
    or self.logitsprocs_need_output_token_ids
    or not thinking_budget_tracks_reqs  # inverted
)
```

When `thinking_budget_tracks_reqs=True` and no penalties are set, the expression evaluates to `False`, so `SamplingMetadata.output_token_ids` is an empty list. `ThinkingBudgetStateHolder.update_state()` iterates zero elements, `_update_think_state()` is never called, `in_end` stays `False`, and `apply_to_logits()` never forces the end token.

## Fix

Remove the erroneous `not` (see PR #41674).

## Impact

- Severity: **High** - `thinking_token_budget` is completely non-functional for typical usage
- Introduced: commit `68dd7db81` (April 29, 2026)

### PR DESCRIPTION
**Pull request by JasonKeyiL**
## Summary

Fixes #41672.

**Root cause:** an inverted boolean condition (`or not thinking_budget_tracks_reqs` instead of `or thinking_budget_tracks_reqs`) in `gpu_input_batch.py`. The condition is logically backwards: it sets `needs_output_token_ids = True` when there are *no* budget-tracking requests, and fails to contribute `True` when there *are* budget-tracking requests.

In practice the bug is currently masked — `self.logitsprocs_need_output_token_ids` is set to `True` at init whenever `reasoning_config is not None` (introduced in the same commit), so `output_token_ids` is never empty and the budget works correctly end-to-end. The fix corrects the latent logical error and removes the unnecessary population of `output_token_ids` when no budget-tracking requests are in the batch.

**Fix:** remove the erroneous `not` (single-token change in `gpu_input_batch.py`).

## Changes

- `vllm/v1/worker/gpu_input_batch.py`: remove `not` from `or not thinking_budget_tracks_reqs`
- `tests/v1/logits_processors/test_correctness.py`: add regression test `test_thinking_budget_enforced_without_penalties` that directly exercises `ThinkingBudgetStateHolder` with an empty vs. live `output_token_ids` list, verifying `in_end` is raised after the budget is exhausted

## Test commands run

```bash
pre-commit run ruff-check --files vllm/v1/worker/gpu_input_batch.py tests/v1/logits_processors/test_correctness.py
pre-commit run ruff-format --files vllm/v1/worker/gpu_input_batch.py tests/v1/logits_processors/test_correctness.py

.venv/bin/python -m pytest tests/v1/logits_processors/test_correctness.py::test_thinking_budget_enforced_without_penalties -v
# PASSED
```

### CODE REVIEW

**Comment by JasonKeyiL** [comment]
Hi @njhill , could you please add the 'verified'  label so CI can run? Thanks!

File: vllm/v1/worker/gpu_input_batch.py [hunk 1]
```diff
@@ -883,7 +876,7 @@ def _make_sampling_metadata(self) -> SamplingMetadata:
             not self.no_penalties
             or bool(self.bad_words_token_ids)
             or self.logitsprocs_need_output_token_ids
```
Comment (rishitdholakia13) [inline vllm/v1/worker/gpu_input_batch.py:886]: Hey @JasonKeyiL , Thanks for catching this and for the fix. I was wondering: with reasoning mode on, ReasoningConfig is not None, so I’d expect` self.logitsprocs_need_output_token_ids` to be set, since that was verified with the E2E test as well. How did we end up with `thinking_budget_tracks_reqs` being set to `True` while `self.logitsprocs_need_output_token_ids` was False?

Comment (JasonKeyiL) [reply vllm/v1/worker/gpu_input_batch.py:886]: Thanks for reply and follow up! @rishitdholakia13 
sorry I missed it, you are right, `self.logitsprocs_need_output_token_ids` is already True at first whenever reasoning_config is not None, `output_token_ids` shouldn't be empty in practice
but isn't the `or not thinking_budget_tracks_reqs`  logically inverted? but given it's masked, is it worth fixing or should I close this?

Comment (rishitdholakia13) [reply vllm/v1/worker/gpu_input_batch.py:886]: Yes, definitely we can merge this fix.

Comment (JasonKeyiL) [reply vllm/v1/worker/gpu_input_batch.py:886]: Thank you, Hi @njhill Nick, would you be able to take a look and put a verify label? thanks!

**Review by njhill — APPROVED** [review]

**Comment by JasonKeyiL** [comment]
@njhill looks like a globe infra issue, CI failed with  `pytorch-compilation-unit-tests, pytorch-fullgraph-smoke-test,`  3 times, could you manually merge it when you get a chance? Thanks you!
